File: ws/first_ws/src/topic_pkg/src/custom_msg_subscriber.cpp
//
// Created by zcb on 2021/4/2.
//

#include <iostream>
#include <ros/ros.h>
#include <demo_msgs/Student.h>

using namespace std;

void topicCallBk(const demo_msgs::Student::ConstPtr & msg){
    cout << "---" << endl;
    cout << "name:"<< msg->name << endl;
    cout << "age:"<< msg->age << endl;
}


int main(int argc,char** argv){
    string nodeName = "subscriber_customer_msg_cpp_node";
    ros::init(argc,argv,nodeName);
    ros::NodeHandle node;


    string topicName = "/custom/msg/topic";
    ros::Subscriber sub = node.subscribe(topicName,1000,topicCallBk);

    ros::spin();

    return 0;
}

File: ros/ws01/first_ws/src/topic_pkg/scripts/custom_msg_subscriber.py
#! /usr/bin/env python
# coding:utf8

import rospy
from demo_msgs.msg import Student


def topic_callback(msg):
    if not isinstance(msg, Student):
        return
    print "name:%s" % msg.name
    print "age:%d" % msg.age


if __name__ == '__main__':
    nodeName = "subscribe_custom_msg_py_node"
    rospy.init_node(nodeName)

    topicName = "/custom/msg/topic"
    rospy.Subscriber(topicName, Student, topic_callback)

    rospy.spin()
